core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -290,22 +287,15 @@ final class ShuffleBlockFetcherIterator(
       sendRequest(fetchRequests.dequeue())
     }
 
-    val iteratorTry: Try[Iterator[Any]] = result match {
+    val iteratorTry: Try[InputStream] = result match {
       case FailureFetchResult(_, e) =>
         Failure(e)
       case SuccessFetchResult(blockId, _, buf) =>
         // There is a chance that createInputStream can fail (e.g. fetching a local file that does
         // not exist, SPARK-4085). In that case, we should propagate the right exception so
         // the scheduler gets a FetchFailedException.
-        Try(buf.createInputStream()).map { is0 =>
-          val is = blockManager.wrapForCompression(blockId, is0)
-          val iter = serializerInstance.deserializeStream(is).asKeyValueIterator
-          CompletionIterator[Any, Iterator[Any]](iter, {
-            // Once the iterator is exhausted, release the buffer and set currentResult to null
-            // so we don't release it again in cleanup.
-            currentResult = null
-            buf.release()
-          })
+        Try(buf.createInputStream()).map { inputStream =>
```

Comment:
We might also treat this patch as an opportunity to revisit why we're using `Try` here.  It might be fine to keep `Try` as the return type but I'm not necessarily convinced that we should be calling `Try.apply()` here since I think it obscures whether we'll need to perform any cleanup after errors (for instance, do we need to free `buf`?  Is `buf` guaranteed to be non-null or could this fail with an NPE on the `buf.createInputStream()` call?  I feel that the `Try.apply()` makes it easy to overlook these concerns.